### PR TITLE
[BEAM-7972] Always use Global window in reshuffle and then apply wind…

### PR DESCRIPTION
**Pull request by angoenka**
…ow again.

**Please** add a meaningful description for your change here

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by angoenka** [comment]
R: @robertwb @y1chi  @tweise

File: sdks/python/apache_beam/transforms/util.py [hunk 1]
```diff
@@ -634,12 +630,13 @@ def restore_timestamps(element, window=DoFn.WindowParam):
...
     ungrouped = pcoll | Map(reify_timestamps)
     ungrouped._windowing = Windowing(
-        window_fn,
+        window.GlobalWindows(),
         triggerfn=AfterCount(1),
         accumulation_mode=AccumulationMode.DISCARDING,
         timestamp_combiner=TimestampCombiner.OUTPUT_AT_EARLIEST)
     result = (ungrouped
               | GroupByKey()
+              | WindowInto(window_fn)
```
Comment (y1chi) [inline sdks/python/apache_beam/transforms/util.py]: seems line 641 is restoring the window, is applying window_fn here makes a difference?

**Comment by angoenka** [comment]
R: @lukecwik

**Comment by y1chi** [comment]
LGTM.
thanks for looking into this.

**Comment by angoenka** [comment]
Gentle reminder for the review.
@robertwb @lukecwik

File: sdks/python/apache_beam/transforms/util.py [hunk 2]
```diff
@@ -612,29 +611,23 @@ def restore_timestamps(element):
...
-          windowing_saved.windowfn.get_window_coder())
-
-      def reify_timestamps(element, timestamp=DoFn.TimestampParam):
+      def reify_timestamps(element,
+                           timestamp=DoFn.TimestampParam,
+                           window=DoFn.WindowParam):
         key, value = element
-        return key, TimestampedValue(value, timestamp)
+        # Transport the window as part of the value and restore it later.
+        return key, TimestampedValue((value, window), timestamp)
```
Comment (robertwb) [inline sdks/python/apache_beam/transforms/util.py]: Any reason not to use a WindowedValue here?

Comment (angoenka) [reply sdks/python/apache_beam/transforms/util.py]: Actually while making the change i realized that the timestamp will be duplicated when using the windowed values without any benefit. Hence dropping the use of windowed value.

**Comment by angoenka** [comment]
Gentle reminder for the review.

File: sdks/python/apache_beam/transforms/util.py (hunk 2 above)
Comment (robertwb) [reply sdks/python/apache_beam/transforms/util.py]: Why would the timestamp be duplicated? The (window, timestamp, value) tuple seems best represented by a WindowedValue. (I would be OK with a plain old 3-tuple as well, but only going half way seems odd).

**Comment by robertwb** [comment]
OK, yuck, that JRH code is really bad (and possibly buggy). The windowing function need not be interpreted by the runner, it should just note it's non-merging and pass things through.

**Comment by angoenka** [comment]
That piece of JRH seems to be very intricate and I would don't think it has an easy fix without a lot of refactoring.
Do you have any better solution in mind?

**Comment by lukecwik** [comment]
> That piece of JRH seems to be very intricate and I would don't think it has an easy fix without a lot of refactoring.
> Do you have any better solution in mind?

The defaulting to global window should be deleted since the Python SDK now does send a proper windowing strategy (same as Go SDK). The code was added as a migration path to allow for differences in where the Python/Go/Java SDKs were when submitting jobs to Dataflow.

**Comment by angoenka** [comment]
Request for another pass for the review.

File: sdks/python/apache_beam/transforms/util.py [hunk 3]
```diff
@@ -612,29 +611,27 @@ def restore_timestamps(element):
...
 
-      def restore_timestamps(element, window=DoFn.WindowParam):
-        # Pass the current window since _IdentityWindowFn wouldn't know how
-        # to generate it.
+      def restore_timestamps(element):
         key, values = element
         return [
             windowed_value.WindowedValue(
-                (key, value.value), value.timestamp, [window])
+                (key, value.value), value.timestamp, value.windows)
```
Comment (robertwb) [inline sdks/python/apache_beam/transforms/util.py]: You could do

```
key, windowed_values = element
return [wv.with_value((key, wv.value)) for wv in windowed_values]
```

File: sdks/python/apache_beam/transforms/util.py [hunk 4]
```diff
@@ -612,29 +611,27 @@ def restore_timestamps(element):
...
             for value in values]
 
     ungrouped = pcoll | Map(reify_timestamps)
+
+    # TODO(BEAM-8104) Using global window as one of the standard window.
+    # This is to mitigate the Java Runner Harness limitation to
```
Comment (robertwb) [inline sdks/python/apache_beam/transforms/util.py]: s/Java Runner Harness/Dataflow Java Runner Harness/

File: sdks/python/apache_beam/transforms/util.py (hunk 3 above)
Comment (angoenka) [reply sdks/python/apache_beam/transforms/util.py]: done

File: sdks/python/apache_beam/transforms/util.py (hunk 4 above)
Comment (angoenka) [reply sdks/python/apache_beam/transforms/util.py]: done

**Comment by robertwb** [comment]
LGTM

**Comment by angoenka** [comment]
Thanks!

**Pull request merged**